File: sdks/sdkwork-im-sdk/sdkwork-im-sdk-typescript/src/transport-selector.ts
/**
 * 传输选择策略实现。
 *
 * 根据客户端运行环境能力，按优先级选择最优传输协议。
 * 支持：
 * - 自动检测当前环境可用的传输（浏览器仅 WebSocket，Node 全部）
 * - 按策略优先级选择，支持自动降级
 * - 手动覆盖（调用方强制指定传输类型）
 * - 从 base URL 推导传输端点
 */

import type {
  ImTransportEndpoint,
  ImTransportFactory,
  ImTransportKind,
  ImTransportSelectionPolicy,
} from './transport.js';
import {
  DEFAULT_TRANSPORT_SELECTION_POLICY,
  parseTransportKindFromUrl,
} from './transport.js';
import { IM_CCP_WEBSOCKET_SUBPROTOCOL } from './ccp-wire.js';
import { IM_REALTIME_WS } from './realtime-api-paths.js';

/**
 * 检测当前环境可用的传输类型列表。
 *
 * 浏览器环境：仅 websocket。
 * Node 环境：websocket + tcp + udp。
 * Tauri/Electron：取决于注入的工厂，至少 websocket。
 */
export function detectAvailableTransports(
  factories: Map<ImTransportKind, ImTransportFactory>,
): ImTransportKind[] {
  const available: ImTransportKind[] = [];
  for (const kind of ['websocket', 'tcp', 'udp'] as const) {
    const factory = factories.get(kind);
    if (factory?.isAvailable()) {
      available.push(kind);
    }
  }
  return available;
}

/**
 * 选择传输工厂。
 *
 * 选择逻辑：
 * 1. 如果指定了 preferred 且该传输可用，直接使用。
 * 2. 否则按 policy.preferred 顺序尝试，选择第一个可用的。
 * 3. 如果 autoFallback=false 且首选不可用，抛出错误。
 * 4. 如果全部不可用，抛出错误。
 *
 * @param factories 传输工厂集合
 * @param policy 选择策略
 * @param preferred 可选的手动覆盖传输类型
 */
export function selectTransportFactory(
  factories: Map<ImTransportKind, ImTransportFactory>,
  policy: ImTransportSelectionPolicy = DEFAULT_TRANSPORT_SELECTION_POLICY,
  preferred?: ImTransportKind,
): ImTransportFactory {
  if (preferred) {
    const factory = factories.get(preferred);
    if (factory?.isAvailable()) {
      return factory;
    }
    if (!policy.autoFallback) {
      throw new Error(
        `Preferred transport "${preferred}" is not available in the current environment.`,
      );
    }
  }

  for (const kind of policy.preferred) {
    const factory = factories.get(kind);
    if (factory?.isAvailable()) {
      return factory;
    }
  }

  throw new Error(
    'No transport is available in the current environment. ' +
      'Provide a custom ImTransportFactory or run in a supported runtime (browser/Node).',
  );
}

/**
 * 从 base URL 和传输类型构建传输端点。
 *
 * - websocket: ws://host:port/im/v3/api/realtime/ws
 * - tcp: tcp://host:port
 * - udp: udp://host:port
 *
 * @param baseUrl 基础 URL（可以是 http/https/ws/wss/tcp/udp 协议）
 * @param kind 传输类型
 * @param deviceId 可选设备 ID（WebSocket 路由绑定用）
 */
export function buildTransportEndpoint(
  baseUrl: string,
  kind: ImTransportKind,
  deviceId?: string,
): ImTransportEndpoint {
  const urlKind = parseTransportKindFromUrl(baseUrl);

  // 如果 URL 已经明确指定了传输类型且与 kind 匹配，直接使用
  if (urlKind === kind && kind !== 'websocket') {
    return {
      kind,
      url: baseUrl,
      deviceId,
    };
  }

  // 根据传输类型转换 URL 协议
  const url = convertBaseUrlForTransport(baseUrl, kind, deviceId);
  return {
    kind,
    url,
    deviceId,
    ...(kind === 'websocket' ? { protocols: [IM_CCP_WEBSOCKET_SUBPROTOCOL] } : {}),
  };
}

/**
 * 将基础 URL 转换为指定传输类型的 URL。
 *
 * - http(s)://host:port → ws(s)://host:port/im/v3/api/realtime/ws （websocket）
 * - http(s)://host:port → tcp://host:port （tcp）
 * - http(s)://host:port → udp://host:port （udp）
 */
function convertBaseUrlForTransport(
  baseUrl: string,
  kind: ImTransportKind,
  deviceId?: string,
): string {
  const trimmed = baseUrl.trim().replace(/\/+$/u, '');

  if (kind === 'websocket') {
    let wsUrl: string;
    if (trimmed.startsWith('https://')) {
      wsUrl = `wss://${trimmed.slice('https://'.length)}`;
    } else if (trimmed.startsWith('http://')) {
      wsUrl = `ws://${trimmed.slice('http://'.length)}`;
    } else if (trimmed.startsWith('wss://') || trimmed.startsWith('ws://')) {
      wsUrl = trimmed;
    } else {
      wsUrl = `ws://${trimmed}`;
    }
    // 追加 realtime ws 路径
    const parsed = new URL(wsUrl);
    const basePath = parsed.pathname.replace(/\/+$/u, '');
    parsed.pathname = basePath.endsWith(IM_REALTIME_WS)
      ? basePath
      : `${basePath}${IM_REALTIME_WS}`;
    if (deviceId) {
      parsed.searchParams.set('deviceId', deviceId);
    }
    return parsed.toString();
  }

  if (kind === 'tcp') {
    const hostPort = extractHostPort(trimmed);
    return `tcp://${hostPort}`;
  }

  if (kind === 'udp') {
    const hostPort = extractHostPort(trimmed);
    return `udp://${hostPort}`;
  }

  return trimmed;
}

/** 从各种协议的 URL 中提取 host:port。 */
function extractHostPort(url: string): string {
  const match = url.match(/^[a-z]+:\/\/([^/]+)/i);
  if (match) {
    return match[1];
  }
  return url;
}

/**
 * 传输选择器，封装完整的传输选择流程。
 */
export class ImTransportSelector {
  private readonly factories: Map<ImTransportKind, ImTransportFactory>;
  private readonly policy: ImTransportSelectionPolicy;

  constructor(
    factories: Map<ImTransportKind, ImTransportFactory>,
    policy: ImTransportSelectionPolicy = DEFAULT_TRANSPORT_SELECTION_POLICY,
  ) {
    this.factories = factories;
    this.policy = policy;
  }

  /** 检测当前环境可用的传输类型。 */
  detectAvailable(): ImTransportKind[] {
    return detectAvailableTransports(this.factories);
  }

  /** 获取指定类型的传输工厂。 */
  getFactory(kind: ImTransportKind): ImTransportFactory | undefined {
    return this.factories.get(kind);
  }

  /** 构建指定类型的传输端点。 */
  buildEndpoint(kind: ImTransportKind, baseUrl: string, deviceId?: string): ImTransportEndpoint {
    return buildTransportEndpoint(baseUrl, kind, deviceId);
  }

  /**
   * 构建候选传输列表，用于连接失败降级。
   *
   * 排序规则：
   * 1. 如果指定了 preferredKind 且可用，放在首位
   * 2. 然后按 policy.preferred 顺序追加其他可用传输（排除已添加的）
   *
   * @param preferredKind 可选的手动覆盖传输类型
   * @returns 候选传输类型列表，按优先级排序
   */
  buildCandidateList(preferredKind?: ImTransportKind): ImTransportKind[] {
    const candidates: ImTransportKind[] = [];
    const seen = new Set<ImTransportKind>();

    const tryAdd = (kind: ImTransportKind): void => {
      if (seen.has(kind)) {
        return;
      }
      const factory = this.factories.get(kind);
      if (factory?.isAvailable()) {
        candidates.push(kind);
        seen.add(kind);
      }
    };

    // 手动覆盖优先
    if (preferredKind) {
      tryAdd(preferredKind);
    }
    // 然后按 policy.preferred 顺序追加
    for (const kind of this.policy.preferred) {
      tryAdd(kind);
    }

    return candidates;
  }

  /**
   * 选择传输并构建端点（同步，仅用于已确定传输类型的场景）。
   *
   * @param baseUrl 基础 URL
   * @param preferredKind 可选的手动覆盖传输类型
   * @param deviceId 可选设备 ID
   */
  select(
    baseUrl: string,
    preferredKind?: ImTransportKind,
    deviceId?: string,
  ): { factory: ImTransportFactory; endpoint: ImTransportEndpoint } {
    const factory = selectTransportFactory(this.factories, this.policy, preferredKind);
    const endpoint = buildTransportEndpoint(baseUrl, factory.kind, deviceId);
    return { factory, endpoint };
  }
}

File: sdks/sdkwork-im-sdk/sdkwork-im-sdk-flutter/composed/im_sdk_composed/lib/src/transport_selector.dart
/// Transport selection policy implementation.
///
/// Mirrors the TypeScript `transport-selector.ts`. Selects the optimal
/// transport based on runtime capabilities, with support for:
/// - Auto-detection of available transports
/// - Priority-based selection with automatic fallback
/// - Manual override (caller forces a specific transport kind)
/// - URL conversion from a base URL to a transport-specific endpoint
library;

import 'transport.dart';
import 'transports/tcp_transport.dart';
import 'transports/udp_transport.dart';
import 'transports/websocket_transport.dart';
import 'ccp_wire.dart';

/// Detects available transport kinds from the given factory map.
///
/// Native Dart/Flutter: websocket + tcp + udp.
/// Browser-like: websocket only (if `dart:io` is unavailable).
List<ImTransportKind> detectAvailableTransports(
  Map<ImTransportKind, ImTransportFactory> factories,
) {
  final available = <ImTransportKind>[];
  for (final kind in [
    ImTransportKind.websocket,
    ImTransportKind.tcp,
    ImTransportKind.udp,
  ]) {
    final factory = factories[kind];
    if (factory?.isAvailable() ?? false) {
      available.add(kind);
    }
  }
  return available;
}

/// Selects a transport factory.
///
/// Selection logic:
/// 1. If [preferred] is specified and available, use it directly.
/// 2. Otherwise, try each kind in [ImTransportSelectionPolicy.preferred] order.
/// 3. If `autoFallback` is false and the preferred kind is unavailable, throw.
/// 4. If all are unavailable, throw.
ImTransportFactory selectTransportFactory(
  Map<ImTransportKind, ImTransportFactory> factories, [
  ImTransportSelectionPolicy policy = defaultTransportSelectionPolicy,
  ImTransportKind? preferred,
]) {
  if (preferred != null) {
    final factory = factories[preferred];
    if (factory != null && factory.isAvailable()) {
      return factory;
    }
    if (!policy.autoFallback) {
      throw StateError(
        'Preferred transport "$preferred" is not available in the current environment.',
      );
    }
  }
  for (final kind in policy.preferred) {
    final factory = factories[kind];
    if (factory != null && factory.isAvailable()) {
      return factory;
    }
  }
  throw StateError(
    'No transport is available in the current environment. '
    'Provide a custom ImTransportFactory or run in a supported runtime.',
  );
}

/// Builds a transport endpoint from a base URL and transport kind.
///
/// - websocket: `ws://host:port/im/v3/api/realtime/ws`
/// - tcp: `tcp://host:port`
/// - udp: `udp://host:port`
ImTransportEndpoint buildTransportEndpoint(
  String baseUrl,
  ImTransportKind kind, {
  String? deviceId,
}) {
  final urlKind = parseTransportKindFromUrl(baseUrl);

  // If the URL already specifies the matching transport kind, use it directly.
  if (urlKind == kind && kind != ImTransportKind.websocket) {
    return ImTransportEndpoint(
      kind: kind,
      url: baseUrl,
      deviceId: deviceId,
      protocols: kind == ImTransportKind.websocket
          ? [imCcpWebSocketSubprotocol]
          : null,
    );
  }

  // Otherwise, convert the base URL to the transport-specific URL.
  final url = _convertBaseUrlForTransport(baseUrl, kind, deviceId);
  return ImTransportEndpoint(
    kind: kind,
    url: url,
    deviceId: deviceId,
    protocols: kind == ImTransportKind.websocket
        ? [imCcpWebSocketSubprotocol]
        : null,
  );
}

String _convertBaseUrlForTransport(
  String baseUrl,
  ImTransportKind kind,
  String? deviceId,
) {
  final trimmed = baseUrl.trim().replaceAll(RegExp(r'/+$'), '');

  if (kind == ImTransportKind.websocket) {
    String wsUrl;
    if (trimmed.startsWith('https://')) {
      wsUrl = 'wss://${trimmed.substring('https://'.length)}';
    } else if (trimmed.startsWith('http://')) {
      wsUrl = 'ws://${trimmed.substring('http://'.length)}';
    } else if (trimmed.startsWith('wss://') || trimmed.startsWith('ws://')) {
      wsUrl = trimmed;
    } else {
      wsUrl = 'ws://$trimmed';
    }
    final parsed = Uri.parse(wsUrl);
    final basePath = parsed.path.replaceAll(RegExp(r'/+$'), '');
    final path = basePath.endsWith(imRealtimeWsPath)
        ? basePath
        : '$basePath$imRealtimeWsPath';
    final query = Map<String, String>.from(parsed.queryParameters);
    if (deviceId != null && deviceId.isNotEmpty) {
      query['deviceId'] = deviceId;
    }
    return parsed
        .replace(path: path, queryParameters: query.isEmpty ? null : query)
        .toString();
  }

  if (kind == ImTransportKind.tcp) {
    return 'tcp://${_extractHostPort(trimmed)}';
  }

  if (kind == ImTransportKind.udp) {
    return 'udp://${_extractHostPort(trimmed)}';
  }

  return trimmed;
}

/// Extracts the `host:port` portion from a URL of any scheme.
String _extractHostPort(String url) {
  final match = RegExp(r'^[a-zA-Z]+://([^/]+)').firstMatch(url);
  if (match != null) {
    return match.group(1)!;
  }
  return url;
}

/// The default set of transport factories for native Dart/Flutter runtimes.
///
/// Includes WebSocket, TCP, and UDP factories. All are backed by `dart:io`.
final Map<ImTransportKind, ImTransportFactory> defaultTransportFactories =
    <ImTransportKind, ImTransportFactory>{
  ImTransportKind.websocket: const ImWebSocketTransportFactory(),
  ImTransportKind.tcp: const ImTcpTransportFactory(),
  ImTransportKind.udp: const ImUdpTransportFactory(),
};

/// Transport selector, encapsulating the full transport selection flow.
class ImTransportSelector {
  ImTransportSelector(
    this.factories, [
    this.policy = defaultTransportSelectionPolicy,
  ]);

  final Map<ImTransportKind, ImTransportFactory> factories;
  final ImTransportSelectionPolicy policy;

  /// Detects available transport kinds in the current environment.
  List<ImTransportKind> detectAvailable() =>
      detectAvailableTransports(factories);

  /// Returns the factory for [kind], or null if not registered.
  ImTransportFactory? getFactory(ImTransportKind kind) => factories[kind];

  /// Builds a prioritized candidate list of available transports.
  ///
  /// If [preferredKind] is specified and available, it is placed first.
  /// Remaining transports are added in policy.preferred order.
  /// Unavailable transports are skipped.
  List<ImTransportKind> buildCandidateList({ImTransportKind? preferredKind}) {
    final candidates = <ImTransportKind>[];
    final seen = <ImTransportKind>{};

    void tryAdd(ImTransportKind kind) {
      if (seen.contains(kind)) return;
      final factory = factories[kind];
      if (factory?.isAvailable() ?? false) {
        candidates.add(kind);
        seen.add(kind);
      }
    }

    if (preferredKind != null) {
      tryAdd(preferredKind);
    }
    for (final kind in policy.preferred) {
      tryAdd(kind);
    }

    return candidates;
  }

  /// Builds a transport endpoint for [kind] from the base URL.
  ImTransportEndpoint buildEndpoint(
    ImTransportKind kind,
    String baseUrl, {
    String? deviceId,
  }) {
    return buildTransportEndpoint(baseUrl, kind, deviceId: deviceId);
  }

  /// Selects a transport and builds the corresponding endpoint.
  ///
  /// Returns a record with the selected [factory] and the constructed
  /// [endpoint].
  ({ImTransportFactory factory, ImTransportEndpoint endpoint}) select(
    String baseUrl, {
    ImTransportKind? preferredKind,
    String? deviceId,
  }) {
    final factory = selectTransportFactory(factories, policy, preferredKind);
    final endpoint = buildTransportEndpoint(
      baseUrl,
      factory.kind,
      deviceId: deviceId,
    );
    return (factory: factory, endpoint: endpoint);
  }
}
